File: Shared/Import/XImportDraftBuilder.swift
import Foundation

struct XImportDraftBuilder: Sendable {
    private let client: any XOEmbedFetching
    private let eventLinkImporter: any EventLinkImporting
    private let fxTwitterClient: any FXTwitterFetching
    private let imageDownloader: any RemoteImageDataFetching

    init(
        client: any XOEmbedFetching = XOEmbedClient(),
        eventLinkImporter: any EventLinkImporting = EventLinkImporter(),
        fxTwitterClient: any FXTwitterFetching = FXTwitterClient(),
        imageDownloader: any RemoteImageDataFetching = RemoteImageDataClient()
    ) {
        self.client = client
        self.eventLinkImporter = eventLinkImporter
        self.fxTwitterClient = fxTwitterClient
        self.imageDownloader = imageDownloader
    }

    func makeDraft(from candidate: URL) async throws -> PendingShareImport {
        guard let normalized = XURLValidator.normalizedPostURL(from: candidate) else {
            throw XOEmbedError.invalidURL
        }

        var draft = PendingShareImport(sourceURL: normalized)
        do {
            let metadata = try await client.fetch(postURL: normalized)
            draft.sourceURL = metadata.canonicalURL
            draft.authorName = metadata.authorName
            draft.postText = metadata.postText
            let contentURLs = metadata.linkedURLs + EventLinkImporter.urls(in: metadata.postText ?? "")
            var details = try await eventLinkImporter.importDetails(from: contentURLs)
            // Announcements often live in the post itself. Text-parsed values
            // only fill fields the linked page could not provide.
            if let postText = metadata.postText,
               let textDetails = JapaneseEventTextParser.parse(postText)
                .details(linkedURL: details?.linkedURL ?? metadata.canonicalURL) {
                details = details.map { $0.fillingMissingFields(from: textDetails) } ?? textDetails
            }
            if let details {
                draft.eventDetails = details
            }
        } catch is CancellationError {
            throw CancellationError()
        } catch {
            draft.warning = String(localized: "share.metadata_warning \(error.localizedDescription)")
        }

        // FxTwitter is only an optional media source. Metadata or image failures
        // must leave the editable oEmbed draft usable.
        do {
            let media = try await fxTwitterClient.fetch(postURL: normalized)
            for imageURL in media.imageURLs {
                do {
                    let data = try await imageDownloader.fetchImage(from: imageURL)
                    if !data.isEmpty {
                        draft.imageData = data
                        break
                    }
                } catch is CancellationError {
                    throw CancellationError()
                } catch {
                    continue
                }
            }
        } catch is CancellationError {
            throw CancellationError()
        } catch {
            // The current import behavior does not depend on media availability.
        }

        // Posts without media can still get a cover from the event page.
        if draft.imageData == nil, let imageURL = draft.eventDetails?.imageURL {
            do {
                let data = try await imageDownloader.fetchImage(from: imageURL)
                if !data.isEmpty {
                    draft.imageData = data
                }
            } catch is CancellationError {
                throw CancellationError()
            } catch {
                // A missing cover image never blocks the import.
            }
        }
        return draft
    }
}

File: OshiLifeAndroid/lib/import/x_import_draft_builder.dart
import 'package:oshilife/import/event_link_importer.dart';
import 'package:oshilife/import/fx_twitter_client.dart';
import 'package:oshilife/import/models/pending_share_import.dart';
import 'package:oshilife/import/remote_image_data_client.dart';
import 'package:oshilife/import/x_oembed_client.dart';
import 'package:oshilife/import/x_url_validator.dart';

/// Port of `Shared/Import/XImportDraftBuilder.swift`.
///
/// Ordering contract (pinned by the ported tests):
/// 1. Validate/normalize the URL — invalid throws before any network.
/// 2. oEmbed metadata + event-link details; ANY failure in this block puts
///    a warning on the draft and leaves it editable.
/// 3. FxTwitter media: first downloadable image wins; per-image failures
///    continue; whole-block failure is silent (no warning).
///
/// iOS composes the warning as the localized `share.metadata_warning %@`
/// string. The import layer stays Flutter-free, so the localized wrapping
/// is injected via [metadataWarningFormatter] by the UI layer; the default
/// carries the error description.
class XImportDraftBuilder {
  XImportDraftBuilder({
    XOEmbedFetching? client,
    EventLinkImporting? eventLinkImporter,
    FXTwitterFetching? fxTwitterClient,
    RemoteImageDataFetching? imageDownloader,
    String Function(Object error)? metadataWarningFormatter,
  }) : _client = client ?? XOEmbedClient(),
       _eventLinkImporter = eventLinkImporter ?? EventLinkImporter(),
       _fxTwitterClient = fxTwitterClient ?? FXTwitterClient(),
       _imageDownloader = imageDownloader ?? RemoteImageDataClient(),
       _metadataWarningFormatter = metadataWarningFormatter ?? _describeError;

  final XOEmbedFetching _client;
  final EventLinkImporting _eventLinkImporter;
  final FXTwitterFetching _fxTwitterClient;
  final RemoteImageDataFetching _imageDownloader;
  final String Function(Object error) _metadataWarningFormatter;

  Future<PendingShareImport> makeDraft(Uri candidate) async {
    final normalized = XUrlValidator.normalizedPostUrl(candidate);
    if (normalized == null) throw const XOEmbedException.invalidUrl();

    final draft = PendingShareImport(sourceUrl: normalized);
    try {
      final metadata = await _client.fetch(normalized);
      draft.sourceUrl = metadata.canonicalUrl;
      draft.authorName = metadata.authorName;
      draft.postText = metadata.postText;
      final contentUrls = [
        ...metadata.linkedUrls,
        ...EventLinkImporter.urls(metadata.postText ?? ''),
      ];
      final details = await _eventLinkImporter.importDetails(contentUrls);
      if (details != null) {
        draft.eventDetails = details;
      }
    } on Object catch (error) {
      draft.warning = _metadataWarningFormatter(error);
    }

    // FxTwitter is only an optional media source. Metadata or image
    // failures must leave the editable oEmbed draft usable.
    try {
      final media = await _fxTwitterClient.fetch(normalized);
      for (final imageUrl in media.imageUrls) {
        try {
          final data = await _imageDownloader.fetchImage(imageUrl);
          if (data.isNotEmpty) {
            draft.imageBytes = data;
            break;
          }
        } on Object {
          continue;
        }
      }
    } on Object {
      // The current import behavior does not depend on media availability.
    }
    return draft;
  }

  static String _describeError(Object error) {
    if (error is FXTwitterException) return error.message;
    return error.toString();
  }
}
